gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/GCImpl.java

```diff
@@ -0,0 +1,203 @@
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.projectnessie.client.StreamingUtil;
+import org.projectnessie.client.api.NessieApiV1;
+import org.projectnessie.error.NessieNotFoundException;
+import org.projectnessie.model.Branch;
+import org.projectnessie.model.CommitMeta;
+import org.projectnessie.model.Content;
+import org.projectnessie.model.ContentKey;
+import org.projectnessie.model.EntriesResponse;
+import org.projectnessie.model.ImmutableDelete;
+import org.projectnessie.model.LogResponse;
+import org.projectnessie.model.Reference;
+import org.projectnessie.model.Tag;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Encapsulates the logic to retrieve all content including their content keys over all commits in
+ * all named-references.
+ */
+final class GCImpl implements GC {
+  private static final Logger LOGGER = LoggerFactory.getLogger(GCImpl.class);
+
+  private final NessieApiV1 api;
+  private final Function<Reference, Instant> cutOffTimestamp;
+
+  GCImpl(NessieApiV1 api, Function<Reference, Instant> cutOffTimestamp) {
+    this.api = api;
+    this.cutOffTimestamp = cutOffTimestamp;
+  }
+
+  @Override
+  public <GC_CONTENT_VALUES extends ExpiredContentValues> GCResult<GC_CONTENT_VALUES> performGC(
+      ContentValuesCollector<GC_CONTENT_VALUES> contentValuesCollector) {
+    // Algorithm for performing the GC to identify the expired contents :
+
+    // Walk through each reference.
+    // For each commit in a reference, collect live table keys or dropped table keys.
+    // For each key, get the content.
+    // If the content's key is in dropped table keys,
+    //     add content to expired set (If the content is not live from previous walks of other
+    // references).
+    // Else if the commit is head commit, add contents to live set.
+    // Else check if the content is expired or not based on cutoff time of this reference.
+    //     Based on that add content to respective sets (live or expired).
+    //
+    // Note: So, If a table is dropped in a reference. Then independent of cutoff time,
+    //     all the snapshots for that table in the reference (that are not live in other reference)
+    // will be
+    //         considered for expiry.
+
+    List<Reference> references = api.getAllReferences().get().getReferences();
+    for (Reference reference : references) {
+      Instant cutOffTimestamp = this.cutOffTimestamp.apply(reference);
+      walkReference(
+          contentValuesCollector,
+          reference,
+          commitMeta ->
+              // If the commit time is older than (think: less than, before) cutoff-time, then
+              // commit is expired.
+              Objects.requireNonNull(commitMeta.getCommitTime()).compareTo(cutOffTimestamp) < 0);
+    }
+    return new GCResult<>(contentValuesCollector.contentValues);
```

Comment:
Given the intention to write this to a table would a List<GcRow> be more suitable for a data structure?

Reply:
instead of just having random order list of contents that are expired, it will be good if group them per content id. hence the map. 
Also in a table, single row maps to one content id.

Reply:
![image](https://user-images.githubusercontent.com/5889404/144544729-eebe1e03-56f7-460f-8932-43ee9b8a2488.png)

Thinking something like this for a table output